src/shopilingus/Shop.java

```diff
@@ -111,29 +103,45 @@ public ProductDetails findProductById(int productId) {
 
     public void addProduct(Product product, double quantity) {
 
-        ProductDetails existingProduct = findProductById(product.id);
+        ProductDetails existingProduct = findProductDetailsById(product.id);
 
-        if(existingProduct != null){
+        if (existingProduct != null) {
             existingProduct.quantity += quantity;
-        }else{
+        } else {
             products.add(new ProductDetails(product, quantity));
         }
 
     }
 
-    public Product getProduct(int productId){
-        for(ProductDetails productDetails : products){
-            if(productDetails.product.id == productId){
-                return productDetails.product;
-            }
+    public Product getProduct(int productId) {
+        ProductDetails productDetails = findProductDetailsById(productId);
+
+        if (productDetails != null) {
+            return productDetails.product;
         }
         return null;
     }
 
-    public void deleteProduct(int productId){
+    public void deleteProduct(int productId) {
+        ProductDetails productToRemove = findProductDetailsById(productId);
 
+        if (productToRemove != null) {
+            products.remove(productToRemove);
         }
     }
 
+    public void updateProduct(int productId, Product modifiedProduct){
+        ProductDetails productDetails = findProductDetailsById(productId);
+        deleteProduct(productId);
```

Comment:
you could add an if (when `productDetails` is not null then proceed, otherwise do nothing)